source/templates/writing-helpers.md

```diff
@@ -1,62 +1,383 @@
+});
+```
 
-```app/helpers/full-name.js
-export default Ember.Handlebars.makeBoundHelper( function(person) {
-  return person.get('firstName') + ' ' + person.get('lastName');
-}, 'firstName', 'lastName');
+You can pass as many named arguments as you'd like. They get added to the
+`namedArgs` argument passed to the function:
+
+```handlebars
+{{my-helper option1="hello" option2="world" option3="goodbye cruel
+world"}}
+```
+
+```app/helpers/my-helper.js
+import Ember from "ember";
+export default Ember.Helper.helper(function(params, namedArgs) {
+  let option1 = namedArgs.option1;
+  let option2 = namedArgs.option2;
+  let option3 = namedArgs.option3;
+
+  console.log(option1); // => "hello"
+  console.log(option2); // => "world"
+  console.log(option3); // => "goodbye cruel world"
+});
+```
+
+You can use JavaScript's destructuring assignment shorthand in this case
+as well to clean up the above code:
+
+```app/helpers/my-helper.js
+import Ember from "ember";
+export default Ember.Helper.helper(function(params, { option1, option2, option3 }) {
+  console.log(option1); // => "hello"
+  console.log(option2); // => "world"
+  console.log(option3); // => "goodbye cruel world"
+});
+```
+
+In sum, arguments are good for passing values:
+
+```handlebars
+{{format-date currentDate}}
 ```
 
-You would use the helper like this:
+Hashes are useful for configuring the behavior of a helper:
 
 ```handlebars
-{{fullName person}}
+{{print-current-date format="YYYY MM DD"}}
 ```
 
-Now, whenever the context's person changes, or when any of the
-_dependent keys_ change, the output will automatically update.
+You can have as many of both as you want, so long as the parameters come
+first:
+
+```handlebars
+{{format-date-and-time date time format="YYYY MM DD h:mm" locale="en"}}
+```
+
+The above example contains two arguments:
+
+* `date`
+* `time`
+
+And two named arguments:
+
+* `format="YYY MM DD h:mm"`
+* `locale="en"`
+
+### Stateful Helpers
+
+By default, helpers are _stateless_. They are passed inputs (parameters
+and a hash), they perform an operation on those inputs, and return a
+single output. They have no side-effects and don't save any information
+that is used on subsequent runs of the function.
+
+In some situations, however, you may need to write a helper that
+interacts with the rest of your application. You can create stateful
+helpers that have access to services in your application, and can
+optionally save state as well.
+
+To create a stateful helper, rather than returning a simple function,
+you should return a subclass of `Ember.Helper`. Helper classes must
+contain a `compute` method that behaves the same as the function passed
+to `Ember.Helper.helper`.
+
+In fact, we can refactor the above stateless helper into a stateful
+helper just by making the function into a `compute` method on the class:
+
+```app/helpers/format-currency.js
+import Ember from "ember";
+
+export default Ember.Helper.extend({
+  compute(params, hash) {
```

Comment:
My take: anything written anew should make full use of ES6, for `1.13` docs we update any sections we touch since we're at it, for `2.0` given the major change everything should be consistent.
The "guide assumpts" in the intro sections ties into this.